File: IntraOp_Image_Registration_Manual.m
% INTRAOP IMAGE REGISTRATION TOOL v1
%
% Description:
% This MATLAB script provides an interactive tool for image registration.
% It allows users to load two images: one with a grid (reference) and one
% without a grid (target). Using the provided interface, users can:
%
% 1. Load and display the grid and no-grid images.
% 2. Select control points to align (register) the no-grid image to the grid image.
% 3. Save the control points, transformed image, and other data for later use.
% 4. Toggle between the grid image, no-grid image, and the transformed image.
%
% The tool dynamically adjusts its layout to fit the figure window size,
% ensuring ease of use on various screen resolutions.
%
% Features:
% - Load images (grid and no-grid).
% - Select control points for registration.
% - Apply a projective transformation to register the no-grid image.
% - Toggle visibility between images.
% - Save all relevant data to a .mat file.
%
% Usage:
% - Run the script and use the buttons to load images, register them, and
%   save the results.
%
% Requirements:
% - MATLAB Image Processing Toolbox for `cpselect` and `imwarp`.
%
% Author:
% David Brang
%
% Copyright (c) 2024 David Brang
% All rights reserved.
%
% License: CC BY-NC 4.0
%
% Version: 1.0
% Date: November 21, 2024
%
% Contact:
% For questions or bug reports, contact David Brang at djbrang@umich.edu

function IntraOp_Image_Registration_Manual()
% Create the main figure
fig = figure('Name', 'Image Registration Tool', 'NumberTitle', 'off', ...
    'WindowState', 'maximized');

% Initialize UserData with placeholders for control points, images, and buttons
fig.UserData.cp1 = [];
fig.UserData.cp2 = [];
fig.UserData.transformed_img1 = []; % Transformed no-grid image
fig.UserData.img1 = []; % No-grid image
fig.UserData.img2 = []; % Grid image
fig.UserData.hNoGrid = []; % Handle for no-grid image
fig.UserData.hGrid = []; % Handle for grid image
fig.UserData.hNoGridTransformed = []; % Handle for transformed no-grid image
fig.UserData.buttonHandles = struct();

% Create buttons and store their handles
fig.UserData.buttonHandles.loadMat = uicontrol('Style', 'pushbutton', 'String', 'Load Previous .mat', ...
    'Callback', @(~, ~) loadMat(fig));

fig.UserData.buttonHandles.loadNoGrid = uicontrol('Style', 'pushbutton', 'String', 'Load No Grid Image', ...
    'Callback', @(~, ~) loadNoGrid(fig));

fig.UserData.buttonHandles.loadGrid = uicontrol('Style', 'pushbutton', 'String', 'Load Grid Image', ...
    'Callback', @(~, ~) loadGrid(fig));

fig.UserData.buttonHandles.toggleTransform = uicontrol('Style', 'togglebutton', 'String', 'Toggle Image', ...
    'Callback', @(src, ~) toggleImages(fig, src));

fig.UserData.buttonHandles.selectControlPoints = uicontrol('Style', 'pushbutton', 'String', 'Select Control Points', ...
    'Callback', @(~, ~) selectControlPoints(fig));

fig.UserData.buttonHandles.saveData = uicontrol('Style', 'pushbutton', 'String', 'Save Data', ...
    'Callback', @(~, ~) saveData(fig));

fig.UserData.buttonHandles.quit = uicontrol('Style', 'pushbutton', 'String', 'Quit', ...
    'Callback', @(~, ~) delete(fig));

% Set the resize function after initialization
fig.SizeChangedFcn = @(src, ~) resizeButtons(src);

% Display placeholder message in axes
hAx = axes(fig, 'Position', [0, 0, 1, 1]); % Fill entire figure
disableDefaultInteractivity(hAx); % Disable all default interactions
axis(hAx, 'off'); % Turn off axis ticks and labels
text(0.5, 0.5, 'Load images or previous .mat to begin.', 'HorizontalAlignment', 'center', ...
    'VerticalAlignment', 'middle', 'FontSize', 36, 'Parent', hAx);

% Callback to resize buttons dynamically
    function resizeButtons(src)
        % Retrieve button handles from UserData
        buttonHandles = src.UserData.buttonHandles;

        % Get updated figure dimensions
        figWidth = src.Position(3);
        figHeight = src.Position(4);

        % Define button dimensions and spacing
        buttonWidth = figWidth * 0.1; % 10% of figure width
        buttonHeight = figHeight * 0.05; % 5% of figure height
        buttonSpacing = figWidth * 0.015; % 1.5% of figure width
        quitButtonWidth = buttonWidth / 2; % Quit button is half the size

        % Update button positions
        set(buttonHandles.loadMat, 'Position', [buttonSpacing, 10, buttonWidth, buttonHeight]);
        set(buttonHandles.loadNoGrid, 'Position', [buttonSpacing * 2 + buttonWidth, 10, buttonWidth, buttonHeight]);
        set(buttonHandles.loadGrid, 'Position', [buttonSpacing * 3 + buttonWidth * 2, 10, buttonWidth, buttonHeight]);
        set(buttonHandles.toggleTransform, 'Position', [buttonSpacing * 4 + buttonWidth * 3, 10, buttonWidth, buttonHeight]);
        set(buttonHandles.selectControlPoints, 'Position', [buttonSpacing * 5 + buttonWidth * 4, 10, buttonWidth, buttonHeight]);
        set(buttonHandles.saveData, 'Position', [buttonSpacing * 6 + buttonWidth * 5, 10, buttonWidth, buttonHeight]);
        set(buttonHandles.quit, 'Position', [figWidth - quitButtonWidth - buttonSpacing, 10, quitButtonWidth, buttonHeight]);
    end

% Function to load no-grid image
    function loadNoGrid(fig)
        [file, path] = uigetfile('*.jpg;*.png;*.tif', 'Select No Grid Image');
        if isequal(file, 0)
            return;
        end
        img1 = imread(fullfile(path, file));
        fig.UserData.img1 = img1;

        % Ensure the axes can handle multiple images
        axes(hAx);
        hold(hAx, 'on');

        % Remove placeholder text if present
        delete(findall(fig, 'Type', 'text', 'Parent', hAx));
        axis(hAx, 'on'); % Enable axes functionality for images

        % Display image
        if isempty(fig.UserData.hNoGrid) || ~isvalid(fig.UserData.hNoGrid)
            % fig.UserData.hNoGrid = imshow(img1, 'Parent', hAx);
            fig.UserData.hNoGrid = imshow(img1, 'Parent', hAx, 'InitialMagnification', 'fit');
        else
            set(fig.UserData.hNoGrid, 'CData', img1, 'Visible', 'on');
        end
        % Hide other images if they exist
        if ~isempty(fig.UserData.hGrid) && isvalid(fig.UserData.hGrid)
            set(fig.UserData.hGrid, 'Visible', 'off');
        end
        if ~isempty(fig.UserData.hNoGridTransformed) && isvalid(fig.UserData.hNoGridTransformed)
            set(fig.UserData.hNoGridTransformed, 'Visible', 'off');
        end
    end

% Function to load grid image
    function loadGrid(fig)
        [file, path] = uigetfile('*.jpg;*.png;*.tif', 'Select Grid Image');
        if isequal(file, 0)
            return;
        end
        img2 = imread(fullfile(path, file));
        fig.UserData.img2 = img2;

        % Ensure the axes can handle multiple images
        axes(hAx);
        hold(hAx, 'on');

        % Remove placeholder text if present
        delete(findall(fig, 'Type', 'text', 'Parent', hAx));
        axis(hAx, 'on'); % Enable axes functionality for images

        % Display image
        if isempty(fig.UserData.hGrid) || ~isvalid(fig.UserData.hGrid)
            %fig.UserData.hGrid = imshow(img2, 'Parent', hAx);
            fig.UserData.hGrid = imshow(img2, 'Parent', hAx, 'InitialMagnification', 'fit');
        else
            set(fig.UserData.hGrid, 'CData', img2, 'Visible', 'on');
        end
        % Hide other images if they exist
        if ~isempty(fig.UserData.hNoGrid) && isvalid(fig.UserData.hNoGrid)
            set(fig.UserData.hNoGrid, 'Visible', 'off');
        end
        if ~isempty(fig.UserData.hNoGridTransformed) && isvalid(fig.UserData.hNoGridTransformed)
            set(fig.UserData.hNoGridTransformed, 'Visible', 'off');
        end
    end

% Function to toggle images
    function toggleImages(fig, src)
        % Retrieve handles
        hNoGrid = fig.UserData.hNoGrid;
        hGrid = fig.UserData.hGrid;
        hNoGridTransformed = fig.UserData.hNoGridTransformed;

        % Check if the transformed image exists and is valid
        if ~isempty(hNoGridTransformed) && isvalid(hNoGridTransformed)
            if src.Value == 1
                % Show transformed image, hide others
                set(hNoGridTransformed, 'Visible', 'on');
                if ~isempty(hNoGrid) && isvalid(hNoGrid)
                    set(hNoGrid, 'Visible', 'off');
                end
                if ~isempty(hGrid) && isvalid(hGrid)
                    set(hGrid, 'Visible', 'off');
                end
            else
                % Show original images, hide transformed
                set(hNoGridTransformed, 'Visible', 'off');
                if ~isempty(hNoGrid) && isvalid(hNoGrid)
                    set(hNoGrid, 'Visible', 'on');
                end
                if ~isempty(hGrid) && isvalid(hGrid)
                    set(hGrid, 'Visible', 'on');
                end
            end
        else
            % If transformed image doesn't exist, toggle between original images
            if src.Value == 1
                if ~isempty(hGrid) && isvalid(hGrid)
                    set(hGrid, 'Visible', 'off');
                end
                if ~isempty(hNoGrid) && isvalid(hNoGrid)
                    set(hNoGrid, 'Visible', 'on');
                end
            else
                if ~isempty(hGrid) && isvalid(hGrid)
                    set(hGrid, 'Visible', 'on');
                end
                if ~isempty(hNoGrid) && isvalid(hNoGrid)
                    set(hNoGrid, 'Visible', 'off');
                end
            end
        end
    end

% Function to select control points and register no-grid image to grid image
    function selectControlPoints(fig)
        if isempty(fig.UserData.img1) || isempty(fig.UserData.img2)
            msgbox('Please load both images before selecting control points.', 'Error', 'error');
            return;
        end

        img1 = fig.UserData.img1; % No-grid image
        img2 = fig.UserData.img2; % Grid image
        cp1 = fig.UserData.cp1;
        cp2 = fig.UserData.cp2;

        if isempty(cp1) || isempty(cp2)
            [cp1, cp2] = cpselect(img1, img2, 'Wait', true);
        else
            [cp1, cp2] = cpselect(img1, img2, cp1, cp2, 'Wait', true);
        end

        if size(cp1, 1) >= 4 && size(cp2, 1) == size(cp1, 1)
            tform = fitgeotform2d(cp1, cp2, 'projective');
            outputView = imref2d(size(img2));
            transformed_img1 = imwarp(img1, tform, 'OutputView', outputView);

            fig.UserData.cp1 = cp1;
            fig.UserData.cp2 = cp2;
            fig.UserData.transformed_img1 = transformed_img1;

            axes(hAx);
            hold(hAx, 'on');
            if isempty(fig.UserData.hNoGridTransformed) || ~isvalid(fig.UserData.hNoGridTransformed)
                fig.UserData.hNoGridTransformed = imshow(transformed_img1, 'Parent', hAx);
            else
                set(fig.UserData.hNoGridTransformed, 'CData', transformed_img1, 'Visible', 'on');
            end

            if ~isempty(fig.UserData.hNoGrid) && isvalid(fig.UserData.hNoGrid)
                set(fig.UserData.hNoGrid, 'Visible', 'off');
            end
            if ~isempty(fig.UserData.hGrid) && isvalid(fig.UserData.hGrid)
                set(fig.UserData.hGrid, 'Visible', 'off');
            end

            %msgbox('No-grid image registered to grid image successfully.', 'Success');
        else
            msgbox('Insufficient control points selected. At least 4 pairs are required.', 'Error', 'error');
        end
    end

% Function to save data
function saveData(fig)
    if isempty(fig.UserData.img1) || isempty(fig.UserData.img2)
        msgbox('No images to save. Please load images first.', 'Error', 'error');
        return;
    end
    
    [file, path] = uiputfile('*.mat', 'Save Data As', 'registration_data.mat');
    if isequal(file, 0)
        return;
    end

    % Save the .mat data
    img1 = fig.UserData.img1;
    img2 = fig.UserData.img2;
    cp1 = fig.UserData.cp1;
    cp2 = fig.UserData.cp2;
    transformed_img1 = fig.UserData.transformed_img1;
    save(fullfile(path, file), 'img1', 'img2', 'cp1', 'cp2', 'transformed_img1');

    % Save the images
    imwrite(fig.UserData.hGrid.CData, fullfile(path, 'Grid.jpg'));
    imwrite(fig.UserData.hNoGrid.CData, fullfile(path, 'NoGrid.jpg'));
    if ~isempty(fig.UserData.hNoGridTransformed)
        imwrite(fig.UserData.hNoGridTransformed.CData, fullfile(path, 'RegisteredNoGrid.jpg'));
    end

    % Create a GIF transitioning between the grid and registered no-grid images
    if ~isempty(fig.UserData.hGrid) && ~isempty(fig.UserData.hNoGridTransformed)
        gridImg = fig.UserData.hGrid.CData;
        transformedImg = fig.UserData.hNoGridTransformed.CData;
        
        % Ensure both images are the same size
        if ~isequal(size(gridImg), size(transformedImg))
            transformedImg = imresize(transformedImg, [size(gridImg, 1), size(gridImg, 2)]);
        end
        
        % Create the animation
        gifFile = fullfile(path, 'Registered_Animation.gif');
        for alpha = linspace(0, 1, 10) % Linearly change alpha in 10 steps
            blendedImg = uint8((1 - alpha) * double(gridImg) + alpha * double(transformedImg));
            [imind, cm] = rgb2ind(blendedImg, 256); % Convert to indexed image
            if alpha == 0
                imwrite(imind, cm, gifFile, 'gif', 'LoopCount', inf, 'DelayTime', 0.1);
            else
                imwrite(imind, cm, gifFile, 'gif', 'WriteMode', 'append', 'DelayTime', 0.1);
            end
        end
    end

    msgbox('Data and images saved successfully.', 'Success');
end

% Function to load previous .mat file
    function loadMat(fig)
        [file, path] = uigetfile('*.mat', 'Select .mat File');
        if isequal(file, 0)
            return;
        end
        data = load(fullfile(path, file), 'img1', 'img2', 'cp1', 'cp2', 'transformed_img1');

        % Update UserData with loaded data
        if isfield(data, 'img1')
            fig.UserData.img1 = data.img1;
        end
        if isfield(data, 'img2')
            fig.UserData.img2 = data.img2;
        end
        if isfield(data, 'cp1')
            fig.UserData.cp1 = data.cp1;
        end
        if isfield(data, 'cp2')
            fig.UserData.cp2 = data.cp2;
        end
        if isfield(data, 'transformed_img1')
            fig.UserData.transformed_img1 = data.transformed_img1;
        end

        % Display loaded images
        axes(fig.CurrentAxes); % Ensure the correct axes is active
        hold(fig.CurrentAxes, 'on');
        if ~isempty(fig.UserData.img1)
            if isempty(fig.UserData.hNoGrid) || ~isvalid(fig.UserData.hNoGrid)
                fig.UserData.hNoGrid = imshow(fig.UserData.img1, 'Parent', fig.CurrentAxes);
            else
                set(fig.UserData.hNoGrid, 'CData', fig.UserData.img1, 'Visible', 'on');
            end
        end
        if ~isempty(fig.UserData.img2)
            if isempty(fig.UserData.hGrid) || ~isvalid(fig.UserData.hGrid)
                fig.UserData.hGrid = imshow(fig.UserData.img2, 'Parent', fig.CurrentAxes);
            else
                set(fig.UserData.hGrid, 'CData', fig.UserData.img2, 'Visible', 'on');
            end
        end
        if ~isempty(fig.UserData.transformed_img1)
            if isempty(fig.UserData.hNoGridTransformed) || ~isvalid(fig.UserData.hNoGridTransformed)
                fig.UserData.hNoGridTransformed = imshow(fig.UserData.transformed_img1, 'Parent', fig.CurrentAxes);
                set(fig.UserData.hNoGridTransformed, 'Visible', 'off');
            else
                set(fig.UserData.hNoGridTransformed, 'CData', fig.UserData.transformed_img1, 'Visible', 'off');
            end
        end
        % Set toggle button state and call toggleImages
        toggleButton = fig.UserData.buttonHandles.toggleTransform;
        set(toggleButton, 'Value', 1); % Set toggle button state to ON
        toggleImages(fig, toggleButton); % Call toggleImages with the updated state

        % Remove placeholder text if present
        delete(findall(fig, 'Type', 'text', 'Parent', hAx));
        %axis(hAx, 'on'); % Enable axes functionality for images

        %msgbox('Data loaded successfully.', 'Success');
    end
end

File: IntraOp_Image_Registration_Manual.py
"""
IntraOp_Image_Registration_Manual.py

Author: David Brang, djbrang@umich.edu
Copyright 2024. University of Michigan. All rights reserved.
License: CC BY-NC 4.0

Purpose:
This script facilitates manual image registration of intraoperative photos through interactive control point selection. Users can define corresponding points on two images to achieve precise alignment.

Dependencies:
- Python 3.7+
- Required libraries: 
  - OpenCV (Install with: `pip install opencv-python`)
  - NumPy (Install with: `pip install numpy`)
  - Matplotlib (for interactive point selection; Install with: `pip install matplotlib`)
  - imageio (for GIF creation): Install with pip install imageio

Inputs: Photo w/o grid, then photo w grid
1. Reference Image: The base image to which the target image will be aligned (e.g., `reference_image.png`). Usually without grid.
2. Target Image: The image to be aligned to the reference image (e.g., `target_image.png`). Usually with grid.

Processes:
Control points selected interactively through the GUI.

Outputs:
- Registered Image: The aligned target image saved as a file (e.g., `registered_image.png`).
- (Optional) Visualization of alignment and control points.

Usage:
1. Install the required libraries.
2. Place the input images in the working directory.
3. Run the script: `python IntraOp_Image_Registration_Manual.py reference_image.png target_image.png`.
"""

import matplotlib.pyplot as plt
import argparse
import cv2
import numpy as np
import imageio
import os
from matplotlib.widgets import Cursor  # Importing Cursor widget
from matplotlib.widgets import Button  # Importing Button widget

# Global variables to store clicked points and line references for both images
points_img1 = []
points_img2 = []
lines_img1 = []
lines_img2 = []
point_history = []  # To keep track of points added
fig = None

# Function to handle mouse clicks on the first image
def onclick_img1(event):
    if event.inaxes == ax1:
        x, y = int(event.xdata), int(event.ydata)
        points_img1.append((x, y))
        point_history.append(('img1', (x, y)))
        # Plot the point and store the Line2D object
        line, = ax1.plot(x, y, 'bo', markersize=4, markerfacecolor='none')  # Note the comma to unpack
        lines_img1.append(line)
        fig.canvas.draw()

# Function to handle mouse clicks on the second image
def onclick_img2(event):
    if event.inaxes == ax2:
        x, y = int(event.xdata), int(event.ydata)
        points_img2.append((x, y))
        point_history.append(('img2', (x, y)))
        # Plot the point and store the Line2D object
        line, = ax2.plot(x, y, 'bo', markersize=4, markerfacecolor='none')
        lines_img2.append(line)
        fig.canvas.draw()

# Function to display images, capture clicks, and provide instructions
def manual_point_selection(img1, img2):
    global fig, ax1, ax2

    # Get the dimensions of the images
    height1, width1 = img1.shape[:2]
    height2, width2 = img2.shape[:2]
    max_height = max(height1, height2)
    total_width = width1 + width2

    # Decide the figure size based on image dimensions to make images as large as possible
    dpi = 100  # Dots per inch
    fig_width = total_width / dpi
    fig_height = max_height / dpi

    fig = plt.figure(figsize=(fig_width, fig_height + 50 / dpi), dpi=dpi)  # Add extra space for the button and title

    # Maximize the figure window
    manager = plt.get_current_fig_manager()
    try:
        manager.window.showMaximized()
    except AttributeError:
        try:
            manager.frame.Maximize(True)
        except AttributeError:
            pass  # Could not maximize the window

    # Set the height fractions for button and images
    button_height = 50 / (fig_height * dpi)  # Convert pixels to fraction
    image_height = 1 - button_height - 0.05  # 0.05 for title

    # Add instructions back in above the images
    fig.suptitle('Click corresponding points on both images. Close the window when done.', fontsize=14, color='blue')

    # Create axes for the two images without any margins
    ax1 = fig.add_axes([0, button_height, width1 / total_width, image_height])
    ax2 = fig.add_axes([width1 / total_width, button_height, width2 / total_width, image_height])

    # Display images
    ax1.imshow(cv2.cvtColor(img1, cv2.COLOR_BGR2RGB))
    ax1.axis('off')  # Remove axes
    ax2.imshow(cv2.cvtColor(img2, cv2.COLOR_BGR2RGB))
    ax2.axis('off')  # Remove axes

    # Add crosshair cursor to both images
    cursor1 = Cursor(ax1, useblit=True, color='blue', linewidth=1)
    cursor2 = Cursor(ax2, useblit=True, color='blue', linewidth=1)

    # Connect the click events
    fig.canvas.mpl_connect('button_press_event', onclick_img1)
    fig.canvas.mpl_connect('button_press_event', onclick_img2)

    # Add Undo Button
    # Define the undo function
    def undo_last_point(event):
        """
        Undo the last point clicked on either image.
        Removes the last point from the appropriate list,
        and updates the plot accordingly.
        """
        if point_history:
            last_image, last_point = point_history.pop()
            if last_image == 'img1':
                points_img1.pop()
                last_line = lines_img1.pop()
                last_line.remove()
            elif last_image == 'img2':
                points_img2.pop()
                last_line = lines_img2.pop()
                last_line.remove()
            fig.canvas.draw()
        else:
            print("No points to undo")

    # Add the Undo button to the figure below the images
    button_ax = fig.add_axes([0.45, 0, 0.1, button_height])
    undo_button = Button(button_ax, 'Undo Last Point')
    undo_button.on_clicked(undo_last_point)

    plt.show()

    return points_img1, points_img2

# Function to apply homography transformation and return the registered image
def register_images(img1, img2, points_img1, points_img2):
    # Ensure the points are in the right format
    pts1 = np.float32(points_img1)
    pts2 = np.float32(points_img2)
    
    # Find the homography matrix based on manual points
    h_matrix, status = cv2.findHomography(pts1, pts2, 0)
    
    # Warp the first image to align with the second one using the homography matrix
    height, width, channels = img2.shape
    registered_img = cv2.warpPerspective(img1, h_matrix, (width, height))
    
    return registered_img

# Function to draw matches vertically using manual points
def drawMatches_vertical(img1, img2, pts1, pts2):
    rows1 = img1.shape[0]
    cols1 = img1.shape[1]
    rows2 = img2.shape[0]
    cols2 = img2.shape[1]
    max_cols = max([cols1, cols2])
    
    out = np.zeros((rows1 + rows2, max_cols, 3), dtype='uint8')
    out[:rows1, :cols1] = img1
    out[rows1:rows1 + rows2, :cols2] = img2

    for pt1, pt2 in zip(pts1, pts2):
        x1, y1 = pt1
        x2, y2 = pt2

        # Adjust y2 coordinate because img2 is below img1
        y2 += rows1

        # Draw circles and lines between matched points
        cv2.circle(out, (int(x1), int(y1)), 4, (0, 255, 0), 1)
        cv2.circle(out, (int(x2), int(y2)), 4, (0, 255, 0), 1)
        cv2.line(out, (int(x1), int(y1)), (int(x2), int(y2)), (255, 0, 0), 1)

    return out

# Function to create transition GIF between two images
def create_transition_gif(img1, img2, output_path, num_frames=10):
    frames = []
    alpha_values = np.linspace(0, 1, num_frames)
    for alpha in alpha_values:
        blended = cv2.addWeighted(img1, 1 - alpha, img2, alpha, 0)
        frames.append(cv2.cvtColor(blended, cv2.COLOR_BGR2RGB))  # Convert BGR to RGB for imageio
    imageio.mimsave(output_path, frames, duration=0.1)

if __name__ == "__main__":
    # Parsing command-line arguments for image paths
    parser = argparse.ArgumentParser(description="Manual landmark selection for image registration.")
    parser.add_argument("image1", help="Path to the first image.")
    parser.add_argument("image2", help="Path to the second image.")
    
    args = parser.parse_args()

    # Load the images
    img1 = cv2.imread(args.image1)
    img2 = cv2.imread(args.image2)

    # Check if images are loaded properly
    if img1 is None:
        print(f"Error loading image 1 from path: {args.image1}")
        exit(1)
    if img2 is None:
        print(f"Error loading image 2 from path: {args.image2}")
        exit(1)

    # Call the manual point selection function
    points_img1, points_img2 = manual_point_selection(img1, img2)
    
    # Register the images using the manually selected points
    if len(points_img1) >= 4 and len(points_img2) >= 4:
        registered_img = register_images(img1, img2, points_img1, points_img2)

        # Draw the matches between the images
        matches_img = drawMatches_vertical(img1, img2, points_img1, points_img2)
        
        # Save the registered image and matches image
        common_dir = os.path.commonpath([args.image1, args.image2])
        output_dir = os.path.join(common_dir, 'Output')
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
        
        # Save input image paths for ease of rerunning
        with open(os.path.join(output_dir, 'input_images.txt'), 'w') as f:
            f.write(f"Image 1: {args.image1}\n")
            f.write(f"Image 2: {args.image2}\n")
        
        # Save input images and output images
        cv2.imwrite(os.path.join(output_dir, 'NoGrid.jpg'), img1)
        cv2.imwrite(os.path.join(output_dir, 'Grid.jpg'), img2)
        cv2.imwrite(os.path.join(output_dir, 'RegisteredImage.jpg'), registered_img)
        cv2.imwrite(os.path.join(output_dir, 'Matches.jpg'), matches_img)

        # Create and save the transition GIF using img2 and registered_img
        gif_output_path = os.path.join(output_dir, 'transition.gif')
        create_transition_gif(img2, registered_img, gif_output_path)

        print(f"Output saved in {output_dir}")
    else:
        print("You need at least 4 points in each image for registration.")
